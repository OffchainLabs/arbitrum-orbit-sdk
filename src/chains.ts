import { defineChain } from 'viem';
import {
  mainnet,
  holesky,
  arbitrum as arbitrumOne,
  arbitrumNova,
  sepolia,
  arbitrumSepolia,
} from 'viem/chains';

const nitroTestnodeL1 = defineChain({
  id: 1_337,
  network: 'Nitro Testnode L1',
  name: 'nitro-testnode-l1',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8545'],
    },
    public: {
      http: ['http://127.0.0.1:8545'],
    },
  },
  testnet: true,
});

const nitroTestnodeL2 = defineChain({
  id: 412_346,
  network: 'Nitro Testnode L2',
  name: 'nitro-testnode-l2',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8547'],
    },
    public: {
      http: ['http://127.0.0.1:8547'],
    },
  },
  testnet: true,
});

// the L3 deployed as part of the nitro-testnode framework uses a custom fee token
const nitroTestnodeL3 = defineChain({
  id: 333333,
  network: 'Nitro Testnode L3',
  name: 'nitro-testnode-l3',
  nativeCurrency: { name: 'AppTestToken', symbol: 'APP', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:3347'],
    },
    public: {
      http: ['http://127.0.0.1:3347'],
    },
  },
  testnet: true,
});

export const chains = [
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
] as const;

export {
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
};
