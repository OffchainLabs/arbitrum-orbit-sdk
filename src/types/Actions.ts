import { Address } from 'viem';
import { Prettify } from './utils';

/**
 * Actions require contract address, but as part of decorators, the address might have been passed already to the decorator.
 *
 * If the address was passed to the decorator, it's now optional (we still allow overrides of the address per action).
 * If the action doesn't have any other parameters beside the contract address, then parameters can either be { contract: address } or void
 */
export type ActionParameters<Args, ContractName extends string, Curried extends boolean> = Prettify<
  Curried extends false
    ? Args & { [key in ContractName]: Address }
    : Args extends Record<string, never>
    ?
        | {
            [key in ContractName]: Address;
          }
        | void
    : Args & {
        [key in ContractName]?: Address;
      }
>;
