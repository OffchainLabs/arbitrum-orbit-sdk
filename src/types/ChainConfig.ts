import { Address } from 'viem';

export type ChainConfigArbitrumParams = {
  EnableArbOS: boolean;
  AllowDebugPrecompiles: boolean;
  InitialArbOSVersion: number;
  InitialChainOwner: Address;
  DataAvailabilityCommittee: boolean;
  GenesisBlockNum: number;
  MaxCodeSize: number;
  MaxInitCodeSize: number;
};

export type ChainConfig = {
  chainId: number;
  homesteadBlock: number;
  daoForkBlock: null;
  daoForkSupport: boolean;
  eip150Block: number;
  eip150Hash: string;
  eip155Block: number;
  eip158Block: number;
  byzantiumBlock: number;
  constantinopleBlock: number;
  petersburgBlock: number;
  istanbulBlock: number;
  muirGlacierBlock: number;
  berlinBlock: number;
  londonBlock: number;
  clique: {
    period: number;
    epoch: number;
  };
  arbitrum: ChainConfigArbitrumParams;
};
