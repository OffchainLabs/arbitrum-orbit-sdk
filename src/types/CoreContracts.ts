import { Address } from 'viem';

export type CoreContracts = {
  rollup: Address;
  nativeToken: Address;
  inbox: Address;
  outbox: Address;
  rollupEventInbox: Address;
  challengeManager: Address;
  adminProxy: Address;
  sequencerInbox: Address;
  bridge: Address;
  upgradeExecutor: Address;
  validatorUtils?: Address;
  validatorWalletCreator: Address;
  deployedAtBlockNumber: number;
};
