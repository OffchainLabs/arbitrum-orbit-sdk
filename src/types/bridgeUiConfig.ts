import { Chain, PublicClient, Transport } from 'viem';

/**
 * Bridge UI configuration information for adding a new test arbitrum chain
 */
export type BridgeUiConfig = {
  /**
   * Basic chain information
   */
  chainInfo: {
    chainName: string;
    chainId: number;
    parentChainId: number;
    rpcUrl: string;
    explorerUrl: string;
    nativeToken?: string;
  };
  /**
   * Core protocol contract addresses
   */
  coreContracts: {
    rollup: string;
    inbox: string;
    outbox: string;
    sequencerInbox: string;
    bridge: string;
    nativeToken?: string;
  };
  /**
   * Token bridge contract addresses for both parent chain (L2) and Orbit chain (L3)
   */
  tokenBridgeContracts: {
    /** Parent chain (L2) contract addresses */
    l2Contracts: {
      customGateway: string;
      multicall: string;
      proxyAdmin: string;
      router: string;
      standardGateway: string;
      weth: string;
      wethGateway: string;
    };
    /** Orbit chain (L3) contract addresses */
    l3Contracts: {
      customGateway: string;
      multicall: string;
      proxyAdmin: string;
      router: string;
      standardGateway: string;
      weth: string;
      wethGateway: string;
    };
  };
};

/**
 * Parameters for retrieving bridge UI configuration
 */
export type GetBridgeUiConfigParams<TChain extends Chain | undefined> = {
  parentChain: TChain;
  deploymentTxHash: `0x${string}`;
  chainName?: string;
  rpcUrl?: string;
  explorerUrl?: string;
  parentChainPublicClient?: PublicClient<Transport, TChain>;
};
