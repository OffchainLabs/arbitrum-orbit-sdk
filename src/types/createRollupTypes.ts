import { Address, GetFunctionArgs } from 'viem';

import { rollupCreator } from '../contracts';

export type CreateRollupFunctionInputs = GetFunctionArgs<
  typeof rollupCreator.abi,
  'createRollup'
>['args'];

type RequiredKeys = 'config' | 'batchPoster' | 'validators';

export type CreateRollupParams = Pick<CreateRollupFunctionInputs[0], RequiredKeys> &
  Partial<Omit<CreateRollupFunctionInputs[0], RequiredKeys>>;

export type WithRollupCreatorAddressOverride<T> = T & {
  /**
   * Specifies a custom address for the RollupCreator. By default, the address will be automatically detected based on the provided chain.
   */
  rollupCreatorAddressOverride?: Address;
};
