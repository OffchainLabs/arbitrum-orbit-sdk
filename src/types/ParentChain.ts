import { Chain, Client, PublicClient, Transport } from 'viem';

import { chains, nitroTestnodeL3 } from '../chains';
import { Prettify } from './utils';

// exclude nitro-testnode L3 from the list of parent chains
export type ParentChain = Exclude<(typeof chains)[number], { id: typeof nitroTestnodeL3.id }>;
export type ParentChainId = ParentChain['id'];

export type ParentChainPublicClient = Prettify<
  Omit<PublicClient, 'chain'> & {
    chain: Prettify<Omit<Chain, 'id'> & { id: ParentChainId }>;
  }
>;

function isValidParentChainId(parentChainId: number | undefined): parentChainId is ParentChainId {
  const ids = chains
    // exclude nitro-testnode L3 from the list of parent chains
    .filter((chain) => chain.id !== nitroTestnodeL3.id)
    .map((chain) => chain.id) as Number[];
  return ids.includes(Number(parentChainId));
}

export function validateParentChain<
  TTransport extends Transport = Transport,
  TChain extends Chain | undefined = Chain | undefined,
>(chainIdOrClient: number | Client<TTransport, TChain>): ParentChainId {
  const chainId = typeof chainIdOrClient === 'number' ? chainIdOrClient : chainIdOrClient.chain?.id;

  if (!isValidParentChainId(chainId)) {
    throw new Error(`Parent chain not supported: ${chainId}`);
  }

  return chainId;
}

export function validateParentChainPublicClient<
  TTransport extends Transport = Transport,
  TChain extends Chain | undefined = Chain | undefined,
>(publicClient: PublicClient<TTransport, TChain>): ParentChainPublicClient {
  const chainId = publicClient.chain?.id;

  if (!isValidParentChainId(chainId)) {
    throw new Error(`Parent chain not supported: ${chainId}`);
  }

  return publicClient as ParentChainPublicClient;
}
