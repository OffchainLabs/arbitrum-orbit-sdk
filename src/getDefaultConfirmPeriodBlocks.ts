import { PublicClient } from 'viem';

import { ParentChainId, validateParentChain } from './types/ParentChain';
import { parentChainIsMainnet } from './parentChainIsMainnet';

export function getDefaultConfirmPeriodBlocks(
  parentChainIdOrPublicClient: ParentChainId | PublicClient,
): bigint {
  const isMainnet = parentChainIsMainnet(validateParentChain(parentChainIdOrPublicClient));

  if (!isMainnet) {
    return 150n;
  }

  return 45_818n;
}
