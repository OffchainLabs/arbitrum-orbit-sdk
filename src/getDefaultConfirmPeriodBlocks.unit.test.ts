import { it, expect } from 'vitest';

import { chains } from './chains';
import { ParentChainId } from './types/ParentChain';
import { getDefaultConfirmPeriodBlocks } from './getDefaultConfirmPeriodBlocks';

it('returns default value for confirmPeriodBlocks based on parent chain', () => {
  expect(
    chains
      .filter((chain) => chain.id !== 333333)
      .reduce((acc, value) => {
        return {
          ...acc,
          // it's ok to cast as we've filtered out 333333 above
          [value.id]: getDefaultConfirmPeriodBlocks(value.id as ParentChainId),
        };
      }, {}),
  ).toMatchSnapshot();
});
