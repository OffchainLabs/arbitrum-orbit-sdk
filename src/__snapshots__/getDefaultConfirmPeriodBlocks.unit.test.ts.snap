// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns default value for confirmPeriodBlocks based on parent chain 1`] = `
{
  "1": 45818n,
  "11155111": 150n,
  "1337": 150n,
  "17000": 150n,
  "412346": 150n,
  "42161": 45818n,
  "421614": 150n,
  "42170": 45818n,
}
`;
